File: sesgen.lua
#!/usr/bin/env lua
--------------------------------------------------------------------------------
--                                                                            --
--  Software Engineering Symbol Generator (Lua 5.1 / 5.2)                     --
--  Harrison Ainsworth / HXA7241 : 2011, 2013                                 --
--                                                                            --
--  http://www.hxa.name/articles/content/software-engineering-symbol_hxa7241  --
--  _2009.html                                                                --
--                                                                            --
--  License: CC0 -- http://creativecommons.org/publicdomain/zero/1.0/         --
--                                                                            --
--------------------------------------------------------------------------------




-- Implementation description.
--
-- Data structure: a sparse 2D array, as a hash-table keyed by coords -- this
-- can hold the cells and allow then to take any form.
--
-- Algorithm: start with one cell, grow 7 more cells, then transcribe the cells
-- into a string. Grow by finding all unoccupied positions adjacent to occupied
-- cells and choosing one.




-- data structure primitive operations -----------------------------------------

function init( x, y )

   return { [coordToKey(x, y)] = makeValue() }

end


function makeValue()

   -- good enough randomness for this purpose
   return math.random( 0, 1 )

end


function coordToKey( x, y )

   -- (carefully avoid negative 0)
   return "" .. (x == 0 and 0 or x) .. " " .. y

end


function keyToCoord( key )

   return string.match( key, "(%S+) (%S+)" )

end




-- algorithmic parts -----------------------------------------------------------

-- undefined -> cells
--
function create()

   -- cells -> undefined (cells mutator)
   --
   function grow( cells )

      -- make set of potential cells valid to grow into
      local potentials = {}
      local count      = 0
      -- step through all cells
      for key, _ in pairs(cells) do
         local x, y = keyToCoord( key )

         -- look left, right, below, above
         for a = 0, 1 do
            for b = -1, 1, 2 do
               local adjacent = coordToKey( x + (a * b), y + ((1-a) * b) )
               -- include if not already occupied in cells
               if not (cells[adjacent] or potentials[adjacent]) then
                  potentials[ adjacent ] = true
                  count = count + 1
               end
            end
         end

      end

      -- choose a potential cell
      local newKey = ""
      do
         local index = math.random( count )
         -- step through until chosen index
         for key, _ in pairs(potentials) do
            if 1 == index then
               newKey = key
               break
            end
            index = index - 1
         end
      end

      -- make actual cell
      cells[ newKey ] = makeValue()

   end

   -- start with one, grow 7 more
   local cells = init( 0, 0 )
   for i = 1, 7 do
      grow( cells )
   end

   return cells

end


-- cells -> string
--
function toString( cells )

   -- get cells rectangular bound
   local min = { x=  math.huge, y=  math.huge }
   local max = { x= -math.huge, y= -math.huge }
   for key, _ in pairs(cells) do
      local x, y = keyToCoord( key )
      min.x, min.y = math.min( min.x, x ), math.min( min.y, y )
      max.x, max.y = math.max( max.x, x ), math.max( max.y, y )
   end

   -- transcribe cells into a block of chars
   local str = ""
   for y = min.y, max.y do
      for x = min.x, max.x do
         local cell = cells[ coordToKey(x, y) ]
         str = str .. (cell and cell or " ")
      end
      str = str .. "\n"
   end

   return "\n" .. str

end




-- execution control -----------------------------------------------------------

-- check if help message needed
if #arg > 0 then

   -- print help message
   print("\n" ..
      "  Software Engineering Symbol Generator (Lua 5.1 / 5.2)\n" ..
      "  Harrison Ainsworth / HXA7241 : 2013-01-26\n" ..
      "  http://www.hxa.name/\n" ..
      "\n" ..
      "Run with no arguments to generate an instance of the 'software " ..
      "engineering\n" .. 
      "symbol' -- eight bits in a four-connected structure.\n" ..
      "See: http://www.hxa.name/articles/content/software-engineering-" ..
      "symbol_hxa7241_2009.html\n")

-- execute
else

   -- (wholly) different every run
   -- (note: Lua table elements are randomly ordered -- this is a second
   -- source of randomness in the program, so setting the same seed will not
   -- produce the same results.)
   math.randomseed( math.floor(os.time()) )

   local cells = create()
   print( toString( cells ) )

end

File: sesgen.ml
#!/usr/bin/env ocaml
(*------------------------------------------------------------------------------

   Software Engineering Symbol Generator (OCaml 4.00)
   Harrison Ainsworth / HXA7241 : 2013

   http://www.hxa.name/articles/content/software-engineering-symbol_hxa7241
   _2009.html

   License: CC0 -- http://creativecommons.org/publicdomain/zero/1.0/

------------------------------------------------------------------------------*)




module List =
struct
   include List

   let rec unfold ?(list = []) fn size = if size <= 0 then
      list else unfold ~list:((fn (size - 1)) :: list) fn (size - 1)
end ;;


module CoordSet =
struct
   include Set.Make(
      struct
         let compare = Pervasives.compare
         type t = int * int
      end
   )

   let rec ofList ?(set = empty) list = if list = [] then
      set else ofList ~set:(add (List.hd list) set) (List.tl list)
end ;;


let rec loop fn i iend product =
   if i < iend then loop fn (i + 1) iend (fn i product) else product ;;




(* entry point -------------------------------------------------------------- *)

(* check if help message needed *)
if (Array.length Sys.argv) > 1 then

   print_endline "\n  \
        Software Engineering Symbol Generator (OCaml 4.00)\n  \
        Harrison Ainsworth / HXA7241 : 2013-01-26\n  \
        http://www.hxa.name/\n\
      \n\
      Run with no arguments to generate an instance of the 'software \
      engineering\n\
      symbol' -- eight bits in a four-connected structure.\n\
      See: http://www.hxa.name/articles/content/software-engineering-symbol_\
      hxa7241_2009.html\n"


(* execute *)
else

   let module C = CoordSet in
   let module L = List in


   (* make cells from randomness *)
   let create seed : (C.elt * string) list =

      (* grow incrementally.
         (precondition: cells empty, potentials singleton) *)
      let rec grow cells potentials i : C.t =
         if i > 0 then
            (* choose new cell *)
            let (x, y) = L.nth (C.elements potentials)
               (Random.int (C.cardinal potentials)) in
            (* update *)
            let cells = C.union cells (C.singleton (x, y)) in
            let potentials =
               let adjacents = C.ofList [(x+1,y); (x-1,y); (x,y+1); (x,y-1)] in
               C.diff (C.union potentials adjacents) cells in
            grow cells potentials (i - 1)
         else
            cells in

      (* associate a bit with each *)
      let color cells : (C.elt * string) list = L.combine (C.elements cells)
         (L.unfold (fun _ -> string_of_int(Random.int 2)) (C.cardinal cells)) in

      let () = Random.set_state seed in
      color (grow C.empty (C.singleton (0,0)) 8) in


   (* convert cells to string *)
   let toString cellBits =

      (* get rectangular bound of cells *)
      let xlo, ylo, xhi, yhi =
         if cellBits <> [] then
            let coords = fst (L.split cellBits) in
            let x0, y0 = L.hd coords in
            let minMax (xlo, ylo, xhi, yhi) (x,y) =
               (min xlo x, min ylo y, max xhi x, max yhi y) in
            L.fold_left minMax (x0, y0, x0, y0) coords
         else
            0, 0, 0, 0 in

      (* transcribe cells into block of chars *)
      let row y str =
         let item x str = str ^ (try (L.assoc (x, y) cellBits)
            with Not_found -> " ") in
         (loop item xlo (xhi +1) str) ^ "\n" in
      loop row ylo (yhi +1) "" in


   (* different every run *)
   let () = Random.self_init () in

   let cellBits = create (Random.get_state ()) in
   print_endline ("\n" ^ (toString cellBits)) ;;
